test/lib/callbacks/StateUpdateRequest-test.js

```diff
@@ -0,0 +1,175 @@
+      const objectUnderTest = new StateUpdateRequest(testClientId, testClientSecret);
+
+      const stub = sinon.stub(rp, 'Request').callsFake(function(options) {
+        const {url, method, body} = options;
+        url.should.equal('https://smartthings/callback');
+        method.should.equal('POST');
+        body.authentication.token.should.equal('aaaa-zzzz');
+        body.authentication.tokenType.should.equal('Bearer');
+        return new Promise((resolve, reject) => {
+          resolve({})
+        })
+      });
+      let refreshCallback = false;
+      objectUnderTest.updateState(testCallbackUrls, textCallbackAuth, testDeviceState, function() {
+        refreshCallback = true;
+      });
+      refreshCallback.should.equal(false);
+      stub.restore();
+    });
+  });
+
+  describe('validTokenRequestWithNoRefreshCallback', function() {
+    it('Should execute proper request ', async function() {
+      const objectUnderTest = new StateUpdateRequest(testClientId, testClientSecret);
+
+      const stub = sinon.stub(rp, 'Request').callsFake(function(options) {
+        const {url, method, body} = options;
+        url.should.equal('https://smartthings/callback');
+        method.should.equal('POST');
+        body.authentication.token.should.equal('aaaa-zzzz');
+        body.authentication.tokenType.should.equal('Bearer');
+        return new Promise((resolve, reject) => {
+          resolve({})
+        })
+      });
+      await objectUnderTest.updateState(testCallbackUrls, textCallbackAuth, testDeviceState);
+      stub.restore();
+    });
+  });
+
+  describe('invalidTokenRequestWithRefresh', function() {
+    it('Should execute proper request, refresh token, and re-issue the request', async function() {
+      const objectUnderTest = new StateUpdateRequest(testClientId, testClientSecret);
+
+      let firstTime = true;
+      let secondCall = false;
+      const stub = sinon.stub(rp, 'Request').callsFake(function(options) {
+        const {url, method, body} = options;
+        if (url === 'https://smartthings/callback') {
+          if (firstTime) {
+            firstTime = false
+            return new Promise((resolve, reject) => {
+              reject({statusCode: 401})
+            })
+          }
+          else {
+            secondCall = true;
+            return new Promise((resolve, reject) => {
+              resolve({})
+            })
+          }
+        }
+        else {
+          url.should.equal('https://smartthings/token');
+          return new Promise((resolve, reject) => {
+            resolve({
+              callbackAuthentication: {
+                accessToken: 'aaaa-xxxx',
+                refreshToken: 'yyyy-cccc'
+              }
+            })
+          })
+        }
+      });
+      let refreshCallback = false;
+      await objectUnderTest.updateState(testCallbackUrls, textCallbackAuth, testDeviceState, function(auth) {
+        refreshCallback = true;
```

Comment:
This has too many branching logic for an unit test.  How about?

const stub = sinon.stub();
const rpStub = sinon.stub(rp, 'Request');
rpStub.onCall(0).rejects({statusCode: 401});
rpStub.onCall(1).resolves({});
rpStub.onCall(2).resolves({
              callbackAuthentication: {
                accessToken: 'aaaa-xxxx',
                refreshToken: 'yyyy-cccc'
              }
            });

await objectUnderTest.updateState(testCallbackUrls, textCallbackAuth, testDeviceState, stub);

const [url, method, body] = rpStub.args[0]; 

const [auth] = stub.args[0];   // this index 1 indicates the call order or 2nd call
auth.should.have.property('accessToken');
auth.should.have.property('refreshToken');
auth.accessToken.should.equal('aaaa-xxxx');
auth.refreshToken.should.equal('yyyy-cccc');

rpStub.callCount.should.equal(3);

sub.restore();

Reply:
This worked for me:
```
const objectUnderTest = new StateUpdateRequest(testClientId, testClientSecret);

const stub = sinon.stub();
const rpStub = sinon.stub(rp, 'Request');
rpStub.onCall(0).rejects({statusCode: 401});
rpStub.onCall(1).resolves({
callbackAuthentication: {
  accessToken: 'aaaa-xxxx',
  refreshToken: 'yyyy-cccc'
}
});
rpStub.onCall(2).resolves({});

await objectUnderTest.updateState(testCallbackUrls, textCallbackAuth, testDeviceState, stub);

const {callbackAuthentication} = await rpStub.returnValues[1];

rpStub.callCount.should.equal(3);
stub.callCount.should.equal(1);
callbackAuthentication.should.have.property('accessToken');
callbackAuthentication.should.have.property('refreshToken');
callbackAuthentication.accessToken.should.equal('aaaa-xxxx');
callbackAuthentication.refreshToken.should.equal('yyyy-cccc');

rpStub.restore();
```